Windows support via AppVeyor

AeroXuk opened the pull request:
This pull request is based on the 'BenK10/libpostal_windows' patch and expanded apon.

I have gone through all the files reviewing the changes and build scripts and making minor changes that will allow this library to be built on Windows via MSYS2 & MingW64.  I have setup AppVeyor which from other Issue logs I saw was a requirement for the libpostal dev team to take on Windows support.

AppVeyor has been configured to build and package the resulting binary into a `.zip` file along with the relevant linking files `libpostal.dll`, `libpostal.lib`, `libpostal.exp` & `libpostal.def`.

The package built by AppVeyor can be uploaded as a Release or linked to via the following URL:
`https://ci.appveyor.com/api/projects/<account>/libpostal/artifacts/libpostal.zip`

The changes I have made build successfully on Travis CI & AppVeyor.

The only program not working on this build is the example console app address_parser. This is due to not having a `termios.h` equivalent. So this step of the build is commented out for the Windows build (it is still compiled for the linux build).

AeroXuk commented:
I've also checked that this compiles on WSL (Windows Subsystem for Linux) without any changes. You don't need to copy the contents of the Windows directory into the Build directory. It builds the same way it would on a Linux box.

albarrentine commented:
This is a good start, thanks for putting it together! Ideally if we're going to support Windows, I'd prefer not to have to declare every function definition in every C file I want to test with LIBPOSTAL_EXPORT.

I'd be more amenable to adding the exports if it's only needed in the public header file `libpostal.h`. Is that possible?

AeroXuk commented:
From what I've seen you need the EXPORT definitions in both the header and source files.

If we are only looking at gcc support, then it is possible to remove all the EXPORT definitions completely.  gcc by default will export all symbols if it detects no explicit ` __declspec(dllexport)` declarations anywhere.

MSVC, however, will not export all symbols by default and requires the explicit use of `__declspec(dllexport)`.  So if MSVC compiler support is a long-term goal export definitions may be needed.

An alternative that I have seen is that CMake has a `WINDOWS_EXPORT_ALL_SYMBOLS` mode, which will automatically add exports to all functions of builds regardless of compiler.  I don't have experience in setting up CMake however.

AeroXuk commented:
Regarding openvenues/pypostal#23, I managed to get that to build on windows with some changes to `setup.py`.  As I don't really know python I'll just summaries the changes I had to make, then hopefully there's a way you can implement windows conditionals.

- Within `libraries=[]`, change `postal` to `libpostal` (windows needs the full name, where as on linux you usually drop the `lib` prefix when linking).
- Remove `-Wno-unused-function` from `extra_compile_args`.
- Changed `#include <libpostal/libpostal.h>` for `#include "libpostal.h"` within `pyexpand.c` & `pyparser.c`.
- Added `libpostal.h`, `libpostal.lib` & `libpostal.dll` from AppVeyor into build path.
- Manually coppied `libpostal.dll` into Python directory (`C:\Program Files\Python36\Lib\site-packages\postal-1.0-py3.6-win-amd64.egg\postal\`).

albarrentine commented:
If there's a default location where the headers and DLL get installed, the `include_dirs=[]` and `library_dirs=[]` args to setup could be set to those location on Windows as well. Defaults are `/usr/local/include` and `/usr/local/lib` on *nix systems, and it helped in most of the binding languages to simply add those defaults explicitly so most users could install in one shot. That would probably work without needing to copy the DLL anywhere or modify `pyexpand.c` and `pyparser.c`.

AeroXuk commented:
Windows doesn't really have a default location for headers and libraries.  The .Net framework has the GAC (Global Assembly Cache), however, that's only for Managed binaries.  `C:\Program Files\` on Windows would be the closest thing to `/usr/local/bin/` on *nix, however, the structure is different, each program has its own folder which it can store its binaries in.  Also, they would not be in the Windows binary search path unless the system environment PATH variable is amended.  The norm for windows would be for a program to bundle all its dependencies into its own folder.

albarrentine commented:
Got it. As far as the C lib, a few questions:

1. In the latest version of this PR, only the public API functions/structs need the conditional DLL export for Windows, and the names of all API functions have to be added to `libpostal.def`. For tests of internal functions, the various `.c` files being tested need to be added to sources in `tests/Makefile.am` but that saves having to add dllexports on internal functions (:100:). Are those the only workflow changes?

2. Any reason why `address_parser_SOURCES` is commented out in the Windows version of the Makefile? That tends to be the most frequently-used client program.

3. If the `libpostal_data` script were ported to a Windows `.bat` (or the original is converted to e.g. Python, which should be present on most systems), and the Makefiles are converted into .vcproj, is there anything about this setup that would prevent us from eventually having a regular MSVC build without needing MSYS2/MingW64? Could that still be added relatively easily?

Lastly, would you be able to add instructions for building on Windows to the README, something similar to what we have in the [installation section](https://github.com/openvenues/libpostal#installation) for Mac/Linux? Should include the prerequisites (and ideally step-by-step instructions on how to get them: some of our users are newer to coding/command-line, and very few are familiar with C), as well as the installation process. Everything else looks good to me.

Thanks again for putting together a PR!

AeroXuk commented:
In repy to your questions:

1. Yes, those should be the only workflow changes.
2. At the moment because of linenoise using `termios.h` (which is not available on Windows) we are unable to compile this.  I can look into how this can be patched to work with Windows.
3. I'm looking into making a windows version of the `libpostal_data` script as I will need a way to update this in my own environment.  I will create a push request for this once I've got something.
MSVC support would still be possible in the future and would depend on some of the changes in the push request. The difficulty would be that some of the headers included currently don't have a Windows alternative. Some projects get around this by declaring local copies of the functions they need from them with some #ifdef'ery to call the relevant system version.

I am putting together something for the README and will commit it soon.

AeroXuk commented:
Ok, no more separate src Makefile for windows now. address_parser now compiles for windows via MSys2/MinGW. I have patched linenoise.c to work with Windows. I've also updated address_parser to use the libpostal API, rather than include modules directly.

The Windows to do list only has make a libpostal_data equivalent on it now.

albarrentine commented on src/linenoise/linenoise.c:
```diff
@@ -946,9 +971,13 @@ static int linenoiseRaw(char *buf, size_t buflen, const char *prompt) {
         }
     } else {
         /* Interactive editing. */
+#ifndef _WIN32
```
Does the address_parser client still work on Windows with all of these conditionals? I didn't write linenoise, but it seems like enableRawMode/disableRawMode are not optional.

AeroXuk replied:
I've confirmed that the Windows build works just like the *nix build. Unfortunately, I don't have valgrind on MSys2.

AeroXuk replied:
I've reduced the amount of #ifdef'ery and amended the commit.

albarrentine commented:
Added some review comments. In short:

- wanted to confirm that the address_parser client still works on Windows with all the ifdefs (inputting addresses, displaying the parse result, reversing through history of the session, should work under valgrind without memory leaks, etc.). Ideally, I'd want to port https://github.com/arangodb/linenoise-ng to C and and use utf8proc, already embedded in this repo, to do the UTF-8 awareness bits.
- C files in libpostal use 4-space indentation instead of tabs
- the headers included in address_parser_cli.c are necessary for the `.print_features` functionality and should be restored.

AeroXuk commented on src/address_parser_cli.c:
```diff
@@ -1,147 +1,130 @@
+	}
+
+	libpostal_teardown();
+	libpostal_teardown_parser();
```
Apologies, I made some edits using Visual Studio this time and it appears to have done some reformatting. I'll amend the commit to remove the tabs.

AeroXuk commented on src/address_parser_cli.c:
```diff
@@ -1,147 +1,130 @@
+
+			cstring_array_destroy(command);
+			goto next_input;
+		} /*else if (string_starts_with(input, ".print_features")) {
```
Ok, I'll rollback this change and try to build this app by including all the source files directly.  I would propose a change that a print_features or debug_mode options flag be included within `libpostal_address_parser_options` in the API so that the app can be simplified.

AeroXuk replied:
I've had another think about this, and I now think it may be better to add an extra function to the API `libpostal_parser_print_features(bool)` to switch this on and off. This way the change won't break any existing language bindings, which might happen if an extra flag was added to `libpostal_address_parser_options`.  The reason for making this part of the API is we can't link `libpostal.c` into the address_parser directly now as the header file tells windows it's a dynamic library.  It also neatens the separation between the library and the address_parser application.  Feel free to suggest whichever way you'd prefer this to be implemented/named. New commit to follow with additional API function.

albarrentine commented:
Looks great! Is this ready to merge?

AeroXuk commented:
Yes, I believe this is ready to merge now.

elyksenyah commented:
Yay! Just to figure out how to get this to work now ...

BenK10 commented:
Nice work, guys!

AeroXuk commented:
The latest successful Master branch build of libpostal.dll can be downloaded from AppVeyor here if you don't want to build it yourself:
https://ci.appveyor.com/api/projects/albarrentine/libpostal/artifacts/libpostal.zip

BenK10 commented:
Can explain this part? I don't quite understand how this works. How is it possible to export the test functions without annotating them with `DLLEXPORT`?:

> In the latest version of this PR, only the public API functions/structs need the conditional DLL export for Windows, and the names of all API functions have to be added to libpostal.def. For tests of internal functions, the various .c files being tested need to be added to sources in tests/Makefile.am but that saves having to add dllexports on internal functions (:100:).

I also noticed that the Makefile now has `-D LIBPOSTAL_EXPORTS ` at `libpostal_la_CFLAGS` and `libscanner_la_CFLAGS`, and that `address_parser_LDADD` now has `libscanner.la` instead of `libpostal.la`. Were these changes necessary to achieve this?

AeroXuk commented:
The test functions aren't exported into the .dll anymore.  The source files needed for the tests have been linked into the test application in the test directory (see test/Makefile.am).

Looking at it now, I believe the following two changes can be made:
- Remove `libscanner.la` from `libpostal_la_LIBADD`.
- Remove `-D LIBPOSTAL_EXPORTS` from `libscanner_la_CFLAGS`.

The `-D LIBPOSTAL_EXPORTS` at `libpostal_la_CFLAGS` and swapping `address_parser_LDADD` to use `libpostal.la` instead of `libscanner.la` were necessary to along with other changes.

albarrentine commented:
@BenK10 yeah, so there are two types of tests:

1. Tests that include only the headers, where we need to link to find the symbols in the shared lib
2. Tests of functions not exposed in the header, which include the whole `.c` file directly, and do not need to be linked

For the first type of test, we were previously linking the symbols from the shared lib, which saved having to compile the source files twice. On Windows that meant everything would have to be dllexported. By adding the various source files to the test Makefile targets, in exchange for having to compile them twice (just added `make -j4` to all builds to compensate), the files are now compiled directly as part of the test program, so only libpostal's public API functions need to be exported.

albarrentine commented:
@AeroXuk `libscanner.la` is a dependency of `libpostal.la` because the generated scanner.c file is very large and has to be compiled without any optimization (`-O0` explicitly) or it will take an unreasonable amount of time to finish (I've left it running for hours and it never completed). The only method I'm aware of for compiling with different optimization levels is to define two targets and link them. Otherwise, scanner.c would need to be added to all the targets and compiled with the same optimization options (usually `-O2` or `-O3`).

linuzer commented:
> The latest successful Master branch build of libpostal.dll can be downloaded from AppVeyor here if you don't want to build it yourself:
https://ci.appveyor.com/api/projects/albarrentine/libpostal/artifacts/libpostal.zip

This does not seem to work, I only get this message:
`{"message":"\"job\" parameter must be specified if build contains multiple jobs.\r\nParameter name: job"}`

Am I doing something fundamental wrong?

AeroXuk commented:
Apologies the url has changed since both 32 and 64 bit builds where added.
These links are for the current 32-bit and 64-bit configurations on AppVeyor:

32-bit: https://ci.appveyor.com/api/projects/albarrentine/libpostal/artifacts/libpostal.zip?branch=master&job=Environment%3A%20COMPILER%3Dmsys2%2C%20PLATFORM%3Dx86%2C%20MSYS2_ARCH%3Dx86%2C%20MSYS2_DIR%3Dmsys64%2C%20MSYSTEM%3DMINGW32%2C%20BIT%3D32

64-bit: https://ci.appveyor.com/api/projects/albarrentine/libpostal/artifacts/libpostal.zip?branch=master&Environment%3A%20COMPILER%3Dmsys2%2C%20PLATFORM%3Dx64%2C%20MSYS2_ARCH%3Dx86_64%2C%20MSYS2_DIR%3Dmsys64%2C%20MSYSTEM%3DMINGW64%2C%20BIT%3D64

linuzer commented:
Thank you very much!

Just a little detail: In the 64-bit link seems like the "job=" part is imssing, so this link work for me:
[https://ci.appveyor.com/api/projects/albarrentine/libpostal/artifacts/libpostal.zip?branch=master&job=Environment%3A%20COMPILER%3Dmsys2%2C%20PLATFORM%3Dx64%2C%20MSYS2_ARCH%3Dx86_64%2C%20MSYS2_DIR%3Dmsys64%2C%20MSYSTEM%3DMINGW64%2C%20BIT%3D64](url)

linuzer commented:
I'm trying to access the DLL from MS VisualStudio. I believe I have correctly referenced and declared the functions, but all the Setup-functions return only FALSE. I read somewhere, that the whole Data-download is not yet converted to Windows-Environment, so I thought I just copy the data-directory from my Linux-environment... is this a valid assumption? If not, how can I prepare the data directory under Windows?
Next thing ist, how am I supposed to provide the DataDir: With Backslash (Like in Windows-World)? Double Backslash? Forward Slash? DriveLetter? Network-Drive (\\Server\Share\Dir)? Terminating Backslash or not?

Maybe someone could give me a little hint...
Thanks a lot!

AeroXuk commented:
The precompiled version of libpostal on windows defaults to `C:\` as libpostal automatically appends `libpostal` on the end.  Resulting in `C:\libpostal\` being the data dir.  I downloaded the data files using the data-download script on both msys2 and within WSL using their respective windows virtual dirs (something like `/mnt/C/`).

I am working on converting the data-download script to a powershell script for Windows, but it's not finished.

linuzer commented:
Thanks a lot, that sounds good!
So my idea of just copying the linux-data dir to c:\libpostal does not work?
I will try the WSL way, just need to update Windows first and install WSL…

AeroXuk commented:
I haven't tested it, I think it may work. WSL is a linux kernal co-running with windows.  So I guess the data dir works from either.

linuzer commented:
Just for reference, I had no success on just copying the datadir from Ubuntu to Windows, but running the whole download/build of libpostal under WSL works. So thanks a lot @AeroXuk and everybody elso for making libpostal available for the Windows-World!

By the way, @mapo80 created here #248 a C# wrapper, so libpostal can now be used very easily from .net-projects! Great work! Thank you everybody a lot, especially @albarrentine for creating libpostal in first place!

AeroXuk commented:
@linuzer Just a word of warning, that C# wrapper doesn't do UTF-8 conversion so can cause libpostal to hang when invalid characters are passed through. I've almost finished a fork of it which makes the API more .NET like and added UTF-8 string conversion. Hope to have it uploaded to GitHub later this evening.

AeroXuk commented:
I've updated my [LibPostalNet fork](https://github.com/AeroXuk/LibPostalNet) including usage example and links to the prerequisites.

AeroXuk commented:
@albarrentine I've had some time to finish the Windows PowerShell script equivalent of libpostal_data. 
Should I put in a Pull request for it? I read somewhere on this repository that you're migrating away from AWS/S3?

albarrentine commented:
@AeroXuk thanks! Yes feel free to make a PR - the plan is just to migrate the URLs to Github but shouldn't change much otherwise since it's all just `curl` with byte offsets.
